File: java/app/src/main/java/com/jossy/android/mobilebenchmarkappjava/utils/BufferedCsvWriter.java
package com.jossy.android.mobilebenchmarkappjava.io;

import android.util.Log;

import com.jossy.android.mobilebenchmarkappjava.data.TestEntry;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Buforowany writer CSV zoptymalizowany pod kątem dużych zbiorów danych (do 1M
 * próbek).
 */
public class BufferedCsvWriter implements AutoCloseable {

    private static final String TAG = "BufferedCsvWriter";

    private final File outputFile;
    private final File checkpointFile;
    private final int bufferCapacity;
    private final int writeBufferBytes;
    private final List<TestEntry> buffer;
    private final ReentrantLock lock = new ReentrantLock();
    private final ExecutorService writeExecutor;
    private final AtomicInteger totalWritten = new AtomicInteger(0);
    private final AtomicInteger totalBuffered = new AtomicInteger(0);

    private BufferedWriter writer;
    private boolean closed = false;
    private long lastFlushTimeMs = System.currentTimeMillis();
    private volatile Throwable lastError = null;

    private final String testName;

    public BufferedCsvWriter(File outputFile, int bufferCapacity, int writeBufferBytes) {
        this(outputFile, bufferCapacity, writeBufferBytes, "unknown");
    }

    public interface CsvFormatter {
        String format(TestEntry entry);

        String getHeader();
    }

    private final CsvFormatter formatter;

    public BufferedCsvWriter(File outputFile, int bufferCapacity, int writeBufferBytes, String testName) {
        this(outputFile, bufferCapacity, writeBufferBytes, testName, null);
    }

    public BufferedCsvWriter(File outputFile, int bufferCapacity, int writeBufferBytes, String testName,
            CsvFormatter formatter) {
        this.outputFile = outputFile;
        this.checkpointFile = new File(outputFile.getParent(),
                outputFile.getName().replace(".csv", "_checkpoint.csv"));
        this.bufferCapacity = bufferCapacity;
        this.writeBufferBytes = writeBufferBytes;
        this.testName = testName;
        this.formatter = formatter != null ? formatter : new DefaultCsvFormatter();
        this.buffer = new ArrayList<>(bufferCapacity);
        this.writeExecutor = Executors.newSingleThreadExecutor(r -> {
            Thread t = new Thread(r, "csv-writer-" + outputFile.getName());
            t.setPriority(Thread.MIN_PRIORITY);
            return t;
        });
    }

    private class DefaultCsvFormatter implements CsvFormatter {
        @Override
        public String format(TestEntry entry) {
            return entry.iteration + "," +
                    entry.result.getExecutionTimeMs() + "," +
                    csvEscape(entry.result.getDetails()) + "," +
                    entry.intervalStartMs + "," +
                    entry.intervalDurationMs + "," +
                    entry.cumulativeTimeMs + "\n";
        }

        @Override
        public String getHeader() {
            return "iteration,execution_time_ms,details,interval_start_ms,interval_duration_ms,cumulative_time_ms\n";
        }
    }

    public void initialize() throws IOException {
        lock.lock();
        try {
            if (writer != null)
                return;
            File parent = outputFile.getParentFile();
            if (parent != null && !parent.exists() && !parent.mkdirs()) {
                throw new IOException("Cannot create directory: " + parent.getAbsolutePath());
            }
            writer = new BufferedWriter(new FileWriter(outputFile), writeBufferBytes);
            writeHeader();
            Log.i(TAG, "Initialized CSV writer: " + outputFile.getName());
        } finally {
            lock.unlock();
        }
    }

    public void write(TestEntry entry) {
        lock.lock();
        try {
            if (closed)
                return;

            buffer.add(entry);
            totalBuffered.incrementAndGet();

            boolean shouldFlush = buffer.size() >= bufferCapacity ||
                    (System.currentTimeMillis() - lastFlushTimeMs > 30_000 && !buffer.isEmpty());

            if (shouldFlush) {
                flushAsync();
            }
        } finally {
            lock.unlock();
        }
    }

    public void flush() {
        lock.lock();
        try {
            flushBufferInternal();
        } catch (IOException e) {
            lastError = e;
            saveCheckpoint();
        } finally {
            lock.unlock();
        }
    }

    private void flushAsync() {
        if (closed)
            return;

        final List<TestEntry> toWrite = new ArrayList<>(buffer);
        buffer.clear();
        lastFlushTimeMs = System.currentTimeMillis();

        if (toWrite.isEmpty())
            return;

        try {
            writeExecutor.submit(() -> {
                lock.lock();
                try {
                    if (writer == null)
                        return;
                    writeEntries(toWrite);
                    totalWritten.addAndGet(toWrite.size());
                    writer.flush();
                } catch (IOException e) {
                    lastError = e;
                    saveCheckpointInternal(toWrite);
                } finally {
                    lock.unlock();
                }
            });
        } catch (Exception e) {
            Log.w(TAG, "Failed to submit async write", e);
        }
    }

    private void flushBufferInternal() throws IOException {
        if (buffer.isEmpty())
            return;
        List<TestEntry> toWrite = new ArrayList<>(buffer);
        buffer.clear();
        writeEntries(toWrite);
        totalWritten.addAndGet(toWrite.size());
        if (writer != null)
            writer.flush();
    }

    private void writeEntries(List<TestEntry> entries) throws IOException {
        if (writer == null)
            throw new IOException("Writer not initialized");
        StringBuilder sb = new StringBuilder(entries.size() * 120);
        for (TestEntry entry : entries) {
            sb.append(formatter.format(entry));
        }
        writer.write(sb.toString());
    }

    private void writeHeader() throws IOException {
        if (writer != null) {
            writer.write(formatter.getHeader());
        }
    }

    public void saveCheckpoint() {
        lock.lock();
        try {
            saveCheckpointInternal(buffer);
        } finally {
            lock.unlock();
        }
    }

    private void saveCheckpointInternal(List<TestEntry> entries) {
        if (entries == null || entries.isEmpty())
            return;
        try (BufferedWriter cpWriter = new BufferedWriter(new FileWriter(checkpointFile, true))) {
            for (TestEntry entry : entries) {
                cpWriter.write(entry.iteration + "," + entry.result.getExecutionTimeMs() + "," +
                        csvEscape(entry.result.getDetails()) + "," + entry.intervalStartMs + "," +
                        entry.intervalDurationMs + "," + entry.cumulativeTimeMs + "\n");
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to save checkpoint", e);
        }
    }

    private String csvEscape(String value) {
        if (value == null)
            return "";
        boolean needsQuote = value.contains(",") || value.contains("\"") || value.contains("\n")
                || value.contains("\r");
        String escaped = value.replace("\"", "\"\"");
        return needsQuote ? ("\"" + escaped + "\"") : escaped;
    }

    @Override
    public void close() {
        boolean alreadyClosed;
        lock.lock();
        try {
            alreadyClosed = closed;
            closed = true;
        } finally {
            lock.unlock();
        }

        if (alreadyClosed)
            return;

        writeExecutor.shutdown();
        try {
            if (!writeExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
                writeExecutor.shutdownNow();
            }
        } catch (InterruptedException e) {
            writeExecutor.shutdownNow();
            Thread.currentThread().interrupt();
        }

        lock.lock();
        try {
            if (!buffer.isEmpty()) {
                try {
                    flushBufferInternal();
                } catch (IOException e) {
                    saveCheckpointInternal(buffer);
                }
            }
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException ignored) {
                } finally {
                    writer = null;
                }
            }
            if (lastError == null && checkpointFile.exists()) {
                checkpointFile.delete();
            }
        } finally {
            lock.unlock();
        }
    }
}

File: kotlin/app/src/main/kotlin/com/jossy/android/mobilebenchmarkappkotlin/utils/BufferedCsvWriter.kt
package com.jossy.android.mobilebenchmarkappkotlin.utils

import android.util.Log
import com.jossy.android.mobilebenchmarkappkotlin.data.TestEntry
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.File
import java.io.Writer

class BufferedCsvWriter(
    private val outputFile: File,
    private val bufferCapacity: Int,
    private val writeBufferBytes: Int,
    private val testName: String = "unknown"
) : AutoCloseable {

    companion object {
        private const val TAG = "BufferedCsvWriter"
    }

    private val checkpointFile =
        File(outputFile.parentFile, outputFile.name.replace(".csv", "_checkpoint.csv"))
    private val buffer = mutableListOf<TestEntry>()

    private val mutex = Mutex()

    // Scope z osobnym wątkiem (Dispatcherem) do zapisu
    private val writeScope =
        CoroutineScope(Dispatchers.IO + SupervisorJob() + CoroutineName("csv-writer"))

    private var totalWritten = 0
    private var writer: Writer? = null
    private var closed = false
    private var lastFlushTimeMs = System.currentTimeMillis()

    fun initialize() {
        runBlocking {
            mutex.withLock {
                if (writer != null) return@withLock

                outputFile.parentFile?.let { parent ->
                    if (!parent.exists() && !parent.mkdirs()) {
                        throw Exception("Cannot create directory: ${parent.absolutePath}")
                    }
                }

                writer = outputFile.bufferedWriter(bufferSize = writeBufferBytes)
                writeHeader()
                Log.i(TAG, "Initialized CSV writer: ${outputFile.name}")
            }
        }
    }

    fun write(entry: TestEntry) {
        writeScope.launch {
            mutex.withLock {
                if (closed) return@withLock

                buffer.add(entry)

                val timeSinceFlush = System.currentTimeMillis() - lastFlushTimeMs
                val shouldFlush = buffer.size >= bufferCapacity ||
                        (timeSinceFlush > 30_000 && buffer.isNotEmpty())

                if (shouldFlush) {
                    flushAsyncInternal()
                }
            }
        }
    }

    fun flush() {
        runBlocking {
            mutex.withLock {
                try {
                    flushBufferToDisk()
                } catch (e: Exception) {
                    saveCheckpoint()
                }
            }
        }
    }

    private fun flushAsyncInternal() {
        if (closed) return

        // Kopiujemy dane
        val toWrite = buffer.toList()
        buffer.clear()
        lastFlushTimeMs = System.currentTimeMillis()

        if (toWrite.isEmpty()) return

        try {
            writeEntries(toWrite)
            totalWritten += toWrite.size
            writer?.flush()
        } catch (e: Exception) {
            saveCheckpointInternal(toWrite)
        }
    }

    private fun flushBufferToDisk() {
        if (buffer.isEmpty()) return
        val toWrite = buffer.toList()
        buffer.clear()
        writeEntries(toWrite)
        totalWritten += toWrite.size
        writer?.flush()
    }

    private fun writeEntries(entries: List<TestEntry>) {
        val currentWriter = writer ?: return
        val csvContent = buildString {
            entries.forEach { entry ->
                append(entry.iteration).append(',')
                append(entry.result.executionTimeMs).append(',')
                append(csvEscape(entry.result.details)).append(',')
                append(entry.intervalStartMs).append(',')
                append(entry.intervalDurationMs).append(',')
                append(entry.cumulativeTimeMs).append('\n')
            }
        }
        currentWriter.write(csvContent)
    }

    private fun writeHeader() {
        writer?.write("iteration,execution_time_ms,details,interval_start_ms,interval_duration_ms,cumulative_time_ms\n")
    }

    private fun saveCheckpoint() {
        saveCheckpointInternal(buffer.toList())
    }

    private fun saveCheckpointInternal(entries: List<TestEntry>) {
        if (entries.isEmpty()) return
        try {
            checkpointFile.appendText(buildString {
                entries.forEach { entry ->
                    append("${entry.iteration},${entry.result.executionTimeMs}," +
                            "${csvEscape(entry.result.details)},${entry.intervalStartMs}," +
                            "${entry.intervalDurationMs},${entry.cumulativeTimeMs}\n")
                }
            })
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save checkpoint", e)
        }
    }

    private fun csvEscape(value: String?): String {
        val v = value ?: return ""
        val needsQuote = v.any { it in ",\"\n\r" }
        return if (needsQuote) "\"${v.replace("\"", "\"\"")}\"" else v
    }

    override fun close() {
        runBlocking {
            mutex.withLock {
                if (closed) return@withLock
                closed = true

                if (buffer.isNotEmpty()) {
                    try {
                        flushBufferToDisk()
                    } catch (e: Exception) {
                        saveCheckpointInternal(buffer.toList())
                    }
                }

                try {
                    writer?.close()
                } catch (ignored: Exception) {
                } finally {
                    writer = null
                }

                if (checkpointFile.exists()) {
                    checkpointFile.delete()
                }
            }
            // Anulujemy scope po wykonaniu wszystkich operacji
            writeScope.cancel()
        }
    }
}